smol/capp/generate/groundstate/upper_bound/solver.py

```diff
@@ -0,0 +1,483 @@
+from typing import List, NamedTuple, Union
+from warnings import warn
+
+import cvxpy as cp
+import numpy as np
+from cvxpy.constraints.constraint import Constraint
+from numpy.typing import ArrayLike
+
+from smol.capp.generate.groundstate.upper_bound.constraints import (
+    get_composition_space_constraints,
+    get_fixed_composition_constraints,
+    get_normalization_constraints,
+)
+from smol.capp.generate.groundstate.upper_bound.objectives import (
+    get_terms_from_chemical_potentials,
+    get_terms_from_decomposition_processor,
+    get_terms_from_ewald_processor,
+    get_terms_from_expansion_processor,
+)
+from smol.capp.generate.groundstate.upper_bound.terms import (
+    get_expression_and_auxiliary_from_terms,
+)
+from smol.capp.generate.groundstate.upper_bound.variables import (
+    get_occupancy_from_variables,
+    get_variables_from_sublattices,
+)
+from smol.moca.ensemble import Ensemble
+from smol.moca.occu_utils import get_dim_ids_table, occu_to_counts
+from smol.moca.processor import (
+    ClusterDecompositionProcessor,
+    ClusterExpansionProcessor,
+    CompositeProcessor,
+    EwaldProcessor,
+)
+
+__author__ = "Fengyu Xie"
+
+
+class ProblemCanonicals(NamedTuple):
+    """CVXpy Canonical objects representing the underlying optimization problem.
+
+    Attributes:
+        problem (cp.Problem):
+            The cvxpy problem to solve.
+        variables (cp.Variable):
+            Variables corresponding to each species on each site to be optimized.
+        variable_indices (cp.Variable):
+            List of variable index corresponding to each active site index and species
+            indices in its site space. Inactive or restricted sites will not have any
+            variable, but be marked as -1 or -2. See groundstate.upper_bound.variables.
+        auxiliary_variables (cp.Variable):
+            Slack variables used to linearize polynomial pseudo-boolean objective
+            terms.
+        indices_in_auxiliary_products (list of lists of int):
+            A list containing the indices of variables whose product equals to the
+            corresponding auxiliary slack variable.
+        objective_function (cp.Expression):
+            Objective function to be minimized. Usually energy per super-cell.
+        constraints (list of cp.Constraint):
+            List of constraints.
+        num_auxiliary_constraints(int):
+            Number of slack linearize constraints. Slack constraints are always
+            the last constraints in the list.
+    """
+
+    problem: cp.Problem
+    variables: cp.Variable
+    # Index of variables on each active site.
+    variable_indices: List[List[int]]
+    auxiliary_variables: cp.Variable
+    indices_in_auxiliary_products: List[List[int]]
+    objective_function: cp.Expression
+    constraints: Union[List[Constraint], None]
+    num_auxiliary_constraints: int
+
+
+class UpperBoundSolver:
```

Comment:
I still think the class with which users interface should have a more intuitive name. I find it more important to hide implementation details because that allows us to extend functionality without breaking backward compatibility or having to write many different classes.

We should chat about this in person to make a final decision.

Reply:
Maybe name it as PeriodicGroundStateSolver is better, but tell more in Doc.

Reply:
Call the lower-bound as AperiodicLowerBoundSolver in the future...

Reply:
Done.